librad/src/meta/entity.rs

```diff
@@ -823,7 +879,80 @@ where
             return Err(UpdateVerificationError::NoPreviousQuorum);
         }
 
-        Ok(self.with_status::<Verified>())
+        Ok(())
+    }
+}
+
+impl<T> Entity<T, Certified>
+where
+    T: Serialize + DeserializeOwned + Clone + EntityInfoExt,
+{
+    /// Given an entity and its previous revision verifies that the update is
+    /// valid
+    ///
+    /// Does the same checks as `Entity::check_update` but returns
+    /// a typed verified value
+    pub fn verify_update(
+        self,
+        previous: &Option<Entity<T, Verified>>,
+    ) -> Result<Entity<T, Verified>, UpdateVerificationError> {
+        self.check_update(previous)
+            .map(|_| self.with_status::<Verified>())
+    }
+}
+
+impl<T> Entity<T, Draft>
+where
+    T: Serialize + DeserializeOwned + Clone + EntityInfoExt,
+{
+    /// Fully verify this entity
+    pub fn check<CACHE>(&self, cache: &mut CACHE) -> Result<(), Error>
+    where
+        CACHE: EntityCache + EntityKeyOwnershipStore,
+    {
+        self.check_signatures()?;
+        self.check_certifiers(cache)?;
+        let revision = self.revision();
+        if revision == 1 {
+            self.check_update(&None)?;
+            Ok(())
+        } else {
+            let urn = self.urn();
+            match cache.last_verified_revision(&urn) {
+                Some(parent) => {
+                    let generic_parent = parent.as_generic_entity();
+                    let generic_self = self.as_generic_entity();
+                    generic_self.check_update(&Some(generic_parent))?;
+                    Ok(())
+                },
+                None => match cache.revision_status(&urn, revision - 1) {
+                    Some(status) => match status {
+                        EntityRevisionStatus::Verified => Ok(()),
+                        _ => Err(Error::CacheError(cache::Error::MissingParentEntity(
+                            urn.to_owned(),
+                            revision,
+                        ))),
+                    },
+                    None => Err(Error::CacheError(cache::Error::MissingParentEntity(
+                        urn.to_owned(),
+                        revision,
+                    ))),
+                },
+            }
+        }
+    }
+}
+
+impl<T> Entity<T, Draft>
+where
+    T: Serialize + DeserializeOwned + Clone + EntityInfoExt,
+{
+    /// Fully verify this entity
+    pub fn verify<CACHE>(self, cache: &mut CACHE) -> Result<Entity<T, Verified>, Error>
```

Comment:
I would suggest `Cache` or `C` for idiomatics

Reply:
Right, this is not a constant, it's a type...